corporate/lib/stripe.py

```diff
@@ -1868,7 +1868,14 @@ def do_update_plan(self, update_plan_request: UpdatePlanRequest) -> None:
         if status is not None:
             if status == CustomerPlan.ACTIVE:
                 assert plan.status < CustomerPlan.LIVE_STATUS_THRESHOLD
-                do_change_plan_status(plan, status)
+                with transaction.atomic():  # nocoverage
+                    # Switch to a different plan was cancelled. We end the next plan
+                    # and set the current one as active.
+                    if plan.status == CustomerPlan.SWITCH_PLAN_TIER_AT_PLAN_END:
+                        next_plan = self.get_next_plan(plan)
+                        assert next_plan is not None
+                        do_change_plan_status(next_plan, CustomerPlan.ENDED)
```

Comment:
What do we end up with as the start/end dates of the cancelled plan? Just curious if it creates something weird with a negative time window or anything.

Reply:
It should be the same as before like `billing_cycle_anchor` would be date we user decided to schedule upgrade and `end_date` would still be what was set using the upgrade script `add_server_to_legacy_plan`. We are only modifing `status` for the legacy plan everywhere now.